lib/internal/modules/esm/translators.js

```diff
@@ -136,8 +139,32 @@ translators.set('json', async function jsonStrategy(url) {
     throw err;
   }
   CJSModule._cache[modulePath] = module;
-  return createDynamicModule(['default'], url, (reflect) => {
+  return createDynamicModule([], ['default'], url, (reflect) => {
     debug(`Parsing JSONModule ${url}`);
     reflect.exports.default.set(module.exports);
   });
 });
+
+// Strategy for loading a wasm module
+translators.set('wasm', async function(url) {
+  const pathname = fileURLToPath(url);
+  const buffer = await readFileAsync(pathname);
+  debug(`Translating WASMModule ${url}`);
+  let compiled;
+  try {
+    compiled = await WebAssembly.compile(buffer);
+  } catch (err) {
+    err.message = pathname + ': ' + err.message;
+    throw err;
+  }
+
+  const imports =
+      WebAssembly.Module.imports(compiled).map(({ module }) => module);
+  const exports = WebAssembly.Module.exports(compiled).map(({ name }) => name);
+
+  return createDynamicModule(imports, exports, url, (reflect) => {
+    const { exports } = new WebAssembly.Instance(compiled, reflect.imports);
```

Comment:
In a cycle-closing edge, will reflect.imports usually be an empty object? If so, this will be basically the right semantics.

Reply:
On the unexecuted cycle edge, `reflect.imports` would contain all the imported module namespaces, with their named exports, but all of those named exports would be undefined. Function exports would be supported in cycles here though just like ES modules.

Reply:
Actually cycles would throw for any accesses to uninitialized let bindings on the namespaces, since these would be in TDZ.

Reply:
To summarize:

* Unexecuted cycle edge with function exports: fine - function is defined
* Unexecuted cycle edge with let binding: throws - TDZ error

Reply:
And specifically these errors are only thrown if the WASM module actually attempts to access the given binding.

Reply:
Right, so in particular, Wasm exports should always act like const bindings, and all imports are accessed during Instantiate (in the ESM evaluate phase). Does this implementation do that?

Reply:
yep! `reflect.imports` is namespace objects so the imports are accessed directly.